external/esp_idf_port/esp32/esp32_flash.c

```diff
@@ -0,0 +1,326 @@
+ *   page -
+ *
+ * Returned Value:
+ *   Page size or negative value on error.  The following errors are reported
+ *   (errno is not set!):
+ *
+ *     EFAULT: On invalid page
+ *     EIO: On unsuccessful erase
+ *     EROFS: On access to write protected area
+ *     EACCES: Insufficient permissions (read/write protected)
+ *     EPERM: If operation is not permitted due to some other constraints
+ *        (i.e. some internal block is not running etc.)
+ *
+ ****************************************************************************/
+
+ssize_t up_progmem_erasepage(size_t page)
+{
+	irqstate_t irqs;
+	int32_t result = 0;
+	result = result;
+
+	if (page >= up_progmem_npages() + start_page()) {
+		return -EFAULT;
+	}
+    /* Disable IRQs while erasing sector */
+	irqs = irqsave();
+
+	/* do erase */
+	result = spi_flash_erase_sector(page);
+
+	/* Restore IRQs */
+	irqrestore(irqs);
+
+	return result == 0 ? CONFIG_ESP32_FLASH_BLOCK_SIZE : -EIO;
+}
+
+/****************************************************************************
+ * Name: up_progmem_write
+ *
+ * Description:
+ *   Program data at given address
+ *
+ *   Note: this function is not limited to single page and nor it requires
+ *   the address be aligned inside the page boundaries.
+ *
+ * Input Parameters:
+ *   addr  - Address with or without flash offset (absolute or aligned to page0)
+ *   buf   - Pointer to buffer
+ *   count - Number of bytes to write *
+ *
+ * Returned Value:
+ *   Bytes written or negative value on error.  The following errors are
+ *   reported (errno is not set!)
+ *
+ *     EINVAL: if count is not aligned with the flash boundaries (i.e.
+ *        some MCU's require per half-word or even word access)
+ *     EFAULT: On invalid address
+ *     EIO: On unsuccessful write
+ *     EROFS: On access to write protected area
+ *     EACCES: Insufficient permissions (read/write protected)
+ *     EPERM: If operation is not permitted due to some other constraints
+ *        (i.e. some internal block is not running etc.)
+ *
+ ****************************************************************************/
+
+ssize_t up_progmem_write(size_t addr, const void *buf, size_t count)
+{
+	int32_t result = 0;
+	ssize_t ret = 0;
+	irqstate_t irqs;
+
+	/* Disable IRQs while erasing sector */
+	irqs = irqsave();
+
+	if (0 != up_progmem_isaddresswritable(addr)) {
+		ret = -EROFS;
+	} else if ((addr & 0xff) != 0) {
+		//! value should be 256 byte alignment
+		ret = -EFAULT;
+	} else if ((count & 0x3) != 0) {
+		ret = -EINVAL;
+	} else {
+		char *adpt_buf = (char *)kmm_malloc(count);
+		if (adpt_buf == NULL) {
+			printf("fail to alloc memory\n");
+			ret = -EPERM;
+		} else {
+			memcpy(adpt_buf, buf, count);
+			result = spi_flash_write(addr, adpt_buf, count);
+			if (result != 0) {
+				ret = -EIO;
+			} else {
+				ret = count;
+			}
+		}
+	}
+
+	/* Restore IRQs */
+	irqrestore(irqs);
+	if (ret < 0) {
+		printf("Errno: %d, addr %d, buf %p, count %d\n", ret, addr, buf, count);

```

Comment:
many printf in this file, plz check

Reply:
rewrite this file, move to arch/xtensa/src/esp32/esp32_flash.c 
and fix many printf.